File: 2024_12/leetcode2024-12-10.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int knightDialer(int n) {
        vector<vector<int>> neibors = {{4,6},{6,8},{7,9},{4,8},{3,9,0},{},{1,7,0},{2,6},{1,3},{2,4}};
        // 定义dp[i][j]表示第i步走到数字j的位置时的总方案数
        vector<vector<int>> dp(n+1, vector<int>(10));
        int modNum = 1000000007;
        for (int i=0; i<10; i++){
            dp[1][i] = 1;
        }
        for (int i=2; i<=n; i++){
            for (int j=0; j<10; j++){
                for (int k : neibors[j]){
                    dp[i][j] = (dp[i][j]+dp[i-1][k])%modNum;
                }
            }
        }
        int res = 0;
        for (int&i : dp[n]){
            res = (res+i)%modNum;
        }
        return res;
    }
};

int main(){
    Solution s;
    int n = 3131;
    cout << s.knightDialer(n) << endl;
}

File: leetcode2024/2024_12/leetcode2024-12-10.py
class Solution:
    def knightDialer(self, n: int) -> int:
        neibors = [[4,6],[6,8],[7,9],[4,8],[3,9,0],[],[1,7,0],[2,6],[1,3],[2,4]]
        # 定义dp[i][j]表示第i步走到数字j的位置时的总方案数
        dp = [[0]*10 for _ in range(n+1)]
        modNum = 1_000_000_007
        for i in range(10):
            dp[1][i] = 1
        for i in range(2,n+1):
            for j in range(10):
                for k in neibors[j]:
                    dp[i][j] = (dp[i][j]+dp[i-1][k])%modNum
        return sum(dp[n]) % modNum

if __name__ == "__main__":
    s = Solution()
    n = 3131
    print(s.knightDialer(n))
